templates/templates.go

```diff
@@ -0,0 +1,511 @@
+			return a + 1
+		},
+		"dec": func(a int) int {
+			return a - 1
+		},
+		"even": func(a int) bool {
+			return a%2 == 0
+		},
+		"incUInt64": func(a uint64) uint64 {
+			return a + 1
+		},
+		"len": func(a string) int {
+			return len(a)
+		},
+		"lenA": func(a []string) int {
+			return len(a)
+		},
+		"lenP": func(a []permission.Permission) int {
+			return len(a)
+		},
+		"lenR": func(a []role.Role) int {
+			return len(a)
+		},
+		"lenU": func(a []user.User) int {
+			return len(a)
+		},
+		"memberAddPermission": func(id int) bool {
+			u, err := t.User.GetUser(context.Background(), user.User{UserID: id})
+			if err != nil {
+				log.Printf("failed to get user for template(getFuncMaps(memberAddPermission)): %+v", err)
+				return false
+			}
+
+			p, err := t.User.GetPermissionsForUser(context.Background(), u)
+			if err != nil {
+				log.Printf("failed to get permissions for user for template(getFuncMaps(memberAddPermission)): %+v", err)
+				return false
+			}
+
+			for _, perm := range p {
+				if perm.Name == permissions.ManageMembersMembersAdd.String() {
+					return true
+				}
+			}
+			return false
+		},
+		"checkPermission": func(id int, p string) bool {
+			return t.permissionsParser(id, p)
+		},
+		"parseHTMLPermissions": func(perms []permission.Permission) template.HTML {
+			var output strings.Builder
+			for _, p := range perms {
+				output.WriteString(fmt.Sprintf("<tr><th>%d</th><td>%s</td><td>%s</td><td>%d</td><td><a href='/internal/permission/%d'>View</a></td></tr>", p.PermissionID, p.Name, p.Description, p.Roles, p.PermissionID))
+			}
+			return template.HTML(output.String())
+		},
+		"parseHTMLPermission": func(p user.PermissionTemplate, userID int) template.HTML {
+			roleAdmin := t.permissionsParser(userID, permissions.ManageMembersGroup.String())
+			var output, roles strings.Builder
+			if len(p.Roles) > 0 {
+				roles.WriteString("Inherited by: <ol>")
+				for _, r := range p.Roles {
+					if roleAdmin {
+						roles.WriteString(fmt.Sprintf("<li style='list-style-type: none;'><span class='tab'></span><a href='/internal/role/%d'>%s</a></li>", r.RoleID, r.Name))
+					} else {
+						roles.WriteString(fmt.Sprintf("<li style='list-style-type: none;'><span class='tab'></span>%s</li>", r.Name))
+					}
+				}
+				roles.WriteString("</ol>")
+			}
+			output.WriteString(fmt.Sprintf("<p>Permission ID: %d<br>Name: %s<br>Description: %s<br><br>%s</p>", p.PermissionID, p.Name, p.Description, roles.String()))
+			return template.HTML(output.String())
+		},
+		"parseHTMLRoles": func(roles []role.Role) template.HTML {
+			var output strings.Builder
+			for _, r := range roles {
+				output.WriteString(fmt.Sprintf("<tr><th>%d</th><td>%s</td><td>%s</td><td>%d</td><td>%d</td><td><a href='/internal/role/%d'>View</a></td></tr>", r.RoleID, r.Name, r.Description, r.Users, r.Permissions, r.RoleID))
+			}
+			return template.HTML(output.String())
+		},
+		"parseHTMLRole": func(r user.RoleTemplate, userID int) template.HTML {
```

Comment:
Names now make more sense `parsexxxIntoHTML` and also we discovered, me and Ben, that `html/template` is actually rather slow with loops, like 300x slower, this method speeds up from a load time of minutes to milliseconds

Reply:
1. This might just be me at this point, but `parse` still implies "convert text into structured data" (eg. "parse JSON", "parse XML") - I'd name these `roleToHTML` or similar. But at this point that's a NITPICK and I'm happy to agree to disagree.
2. The performance point is odd and slightly concerning though - do you have a benchmark or a link to the slow code so I can test it?

Reply:
I'll find the commit where this is added first and send it to you

Reply:
Summarising Slack conversation: I'm not surprised that Go templates are slower than hand-written code, but the degree of the slowdown is very surprising and likely not caused by `text/template` itself but by slow calls in the templates.

I looked through the templates at https://github.com/ystv/web-auth/blob/780254772edaa1c36cda852673f7ebd0c8c5d357/public/templates/users.tmpl (before the conversion) and found a few issues (e.g. the calls to `checkPermission` in each row each firing off two DB calls), there's probably more issues in that nature.

For now I'm happy to keep the hand-rolled solution as long as it's tidied up slightly (e.g. breaking the long `fmt.Sprintf` calls across multiple lines), but this needs revisiting.

Reply:
I have made it much nicer in my latest commit @markspolakovs can you have a look over it again

Reply:
The checkPermission will come probably tomorrow

Reply:
I have fixed this now, it makes a permission get once instead of up to ~10 times